projects/hathor-wallet-headless/hsm-integration.md

```diff
@@ -0,0 +1,256 @@
+This operation should be done via script, as it is the most critical operation on the HSM within this scope.
+
+As such, a new `scripts/createHSMKey.js` file will be created, offering a way to quickly generate a new BIP32 wallet on a developer machine. This created key name should be inserted on the `config.js` file along with a `walletId` for it. Other wallet generation possibilities are discussed on the _Alternatives_ section of this document.
+
+## Error handling
+Errors while communicating with the HSM are to be treated in such a way that it's clear to the end user that the scope of the problem is related to the HSM communication.
+
+## Wallet Lib integration
+No change on the lib code will be necessary.
+
+The interaction with the Wallet Lib will be similar to the current *Hardware Wallet* implementation on the Wallet Desktop.
+
+The wallet storage will be initialized with the `xPub` ( reference: [`initHWStorage`](https://github.com/HathorNetwork/hathor-wallet/blob/ef57015015375a477cffd72baf62f4e14baf541a/src/storage.js#L127-L140) function ) obtained from the HSM: a read-only wallet. On the 2nd phase, signatures are made with the [`ledger/sendTx()` function](https://github.com/HathorNetwork/hathor-wallet/blob/ef57015015375a477cffd72baf62f4e14baf541a/src/utils/ledger.js#L126-L166), as described on the Headless section above.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Multiple HSM devices
+The proposed structure enforces that only one HSM device can be accessed by a headless application environment, since the authentication credentials are stored on config variables.
+
+Another approach to this would be to transfer these credentials to the `/hsm/start` requests, allowing each `walletId` to store its own credentials besides the HSM key itself.
+
+The downside of this solution is that the security of these credentials would be in responsibility of the end user, instead of wrapped within the security of the Headless Wallet configurations.
+
+# Prior art
+[prior-art]: #prior-art
+
+A similar solution to this integration was implemented on the Desktop Wallet for interacting with the Ledger Hardware Wallet. This design uses many of the concepts and code implementations to achieve a similar objective.
+
+The main difference between the Ledger and the HSM is that here there is no need for human intervention on the steps.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+The only question left is if our application is able to correctly sign a transaction with the HSM in its current state, or if changes will have to be made to the Hathor Lib or the Dinamo-Hsm lib to allow this integration. The 2nd phase will be dependent on this answer, and it is being discussed on [internal issue 205](https://github.com/HathorNetwork/internal-issues/issues/205).
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Full integration on all headless endpoints
```

Comment:
I understand that full compatibility is not the goal of this project but I think we should at least describe what the options of implementations are. This way the discussion of which implementation is better will be faster.

What do you think?

Reply:
Good idea. I added a draft proposal to adapt the `prepareTransaction` with this purpose on [36c319b](https://github.com/HathorNetwork/rfcs/pull/67/commits/36c319bb42dc11b8f853d2d549bc04ebf837eb87).